### ISSUE TITLE
Tray icon is not displayed in KDE (5.4.2)

### ISSUE THREAD

**Issue by ComBin** [jitsi/jitsi#192]
Without plasma-systray-legacy (xembed-sni-proxy) icon is not displayed, and with it is displayed but does not respond to clicks.

**Comment by ibauersachs** [comment]
We use Java for the systray in a platform-independent manner. I don't think it's our responsibility to follow whatever new trend a desktop env sets for displaying tray icons and not bringing the Java package up to date.
But given that we also have problems with other platforms I'm leaving this issue open.

**Comment by hoijui** [comment]
The same problem happens for me on Trisquel 7, default desktop.

As is described in the JavaDoc for the SystemTray class (https://docs.oracle.com/javase/8/docs/api/java/awt/SystemTray.html),
jitsi should first check if the system supports/has a system-tray, and if not, use a workaround. for example, just directly close jitsi (after maybe a warning) when cicking on the X(==close window button of the window manager), and make sure the main window is always open as long as jitsi is running, as in: no minimization to invisible/non-existent system-tray.
in other words: it is not save to assume that all systems will have a system-tray, even if in some case it might actually be a problem betwen Java and the OS/WM, because on some it is clearly not.

#### Trisquel 7 workaround
1. right-click on the start panel
2. click on "Add to Panel ..."
3. choose "Notification Area"
4. click on the button "Add"

after this, jitsi starts fine, and a tray icon is visible.

related ubuntu issue: https://bugs.launchpad.net/ubuntu/+source/tasque/+bug/1448512

jitsi output: https://pastebin.mozilla.org/8854062

**Comment by Starkmann** [comment]
I have this problem with wingpanel under elementaryOs too.
Maybe it is related to this:
https://elementaryos.stackexchange.com/questions/4226/how-can-i-get-applications-to-display-a-system-tray-icon

Without systray icon jitsi is no usable under elementaryOS

**Comment by vindicatorr** [comment]
Bug report submitted to KDE: https://bugs.kde.org/show_bug.cgi?id=362941
I'm just not sure where the responsibility lies. Should Oracle be notified?

Log:
```
.../log/jitsi0.log.0:2016-05-11 07:21:46.177 SEVERE: [278] impl.osdependent.jdic.SystrayServiceJdicImpl.checkInitialized().475 Systray not init
```
And when we are stuck in this position, what config file and value should be changed so it doesn't get stuck in the tray?
I'm guessing "net.java.sip.communicator.impl.systray.showApplication=false" should become "true" for ~/.jitsi/sip-communicator.properties
It opened back up for me (after I killed the java process and restarted). You just have to be sure not to "X" close, but rather File->Quit.

**Comment by ibauersachs** [comment]
Last thing I've read is that legacy systray support is coming back in KDE 5.5. But ultimately there should be better support in Java.

**Comment by github-mweston** [comment]
With KDE Plasma 5.5.5, the situation is slightly better, yet **_very far_** from fixed.

Now the Jitsi icon does appear in the system tray, yet it behaves incorrectly:
On Hover: "JavaEmbeddedFrame" is displayed
On Right Mouse Click: No Action
On Left Mouse Click: No Action
On Left Mouse Double Click: No Action

If the Jitsi window is hidden using the X window button, there is no way to restore Jitsi.

I represent a small community of people who use Jitsi under KDE Plasma 5, and this is a **very** significant problem for us.  We would **_very greatly_** appreciate if this problem was resolved as soon as possible.

**Comment by vindicatorr** [comment]
I'm thinking you probably didn't need to copy/paste what I said in the bug report to KDE which I had previously linked to here.

But while I'm here, I'll go ahead and state I'm using KDE Plasma 5.6.4 and the same jitsi-nightly 2.9.20160414-1 with the same results as before.

"Legacy" is a word I feel has a negative connotation. A bit like deprecated. If something is newer and replacing the legacy(deprecated) feature, then that direction is where things need to be headed. It's sort of why we move on to SSD and not still using 5.25" floppies.
But on the flipside, it's understandable being on the leading edge means perhaps going in one direction, figuring out that it isn't a good direction and instead taking a different one. That would make deciding when to go with something "newer" a timing issue. Again, maybe similar to going the direction of blu-ray and not hd-dvd.

**Comment by hoijui** [comment]
i have patches mostly ready for this, as described in an other issue.. just cleaning them up a bit.

**Comment by ibauersachs** [comment]
@hoijui Thanks! I was fiddling around with the tray this weekend and changed the API quite a bit. You might want to wait with the cleanup until I push these changes.

**Comment by hoijui** [comment]
ahh thanks for the heads-up @ibauersachs!
i will wait then.

**Comment by ibauersachs** [comment]
@hoijui I pushed the changes yesterday. Anything in your patches that still should get in?

**Comment by hoijui** [comment]
@ibauersachs i just tried current master...
i have no (compatible) systray on my machine. when i click on the [X] (among the 3 window buttons), the main window disapears but jitsi keeps running. i also see this on the console:
impl.osdependent.jdic.SystrayServiceJdicImpl.checkInitialized().453 Systray not init
is this expected? do you plan to change that yourself?

**Comment by ibauersachs** [comment]
Hmm. The log entry is normal, but Jitsi shouldn't disappear anymore when clicking on X.

**Comment by ibauersachs** [comment]
Also, any recommendation for a distro to test these problems? And by that, I mean a stock-distro where I don't need to install any additional packages after the main setup (except Jitsi of course).

**Comment by hoijui** [comment]
trisquel, but that only suports free software (including firmware and drivers), so no ATI or NVIDIA graphics are supported, for example.

**Comment by ibauersachs** [comment]
Not installable on Hyper-V (says there's no disk space and no internet connection). Other suggestions?
(The Hyper-V drivers are in the Kernel for ages now, even Ubuntu 12 had them...)

**Comment by hoijui** [comment]
if they are in the kernel as binary blobs only, trisquel will not have them. :/
but maybe it is enough if you just hard-code an exception into the code, where you expect it to happen on a system without a systray, and see what happens on your machine. you can also push fixes to a side-branch, and i test them on my system afterwards, if you do not want to clutter the master branch.

**Comment by ibauersachs** [comment]
They're not blobs, just pure plain free GPL source. Anyway, since 7126030f9835f3c932e28b2f391794af411f3c82, SystrayServiceJdicImpl (or rather getSystemTray) checks the isEnabled property. If that returns true, there's not much we can do - except for doing the new, non-X11 systray stuff through JNA. I found a library that we might be able to use, but I'm not sure when I'll have time to work on that.

**Comment by hoijui** [comment]
@ibauersachs i  pushed my patches (on the old systray interface, before your recent changes) here:
https://github.com/hoijui/jitsi/tree/systrayNone
according to my testing, they work as they should, as we discussed it a few months back on an other issue. i did not test them with a working systray; there should be no change there.

**Comment by ComBin** [comment]
Hi guys, i found intresting thread in in Debian maillist: https://lists.debian.org/debian-kde/2015/09/msg00171.html. May be this somehow to help you.

**Comment by hoijui** [comment]
i see no new info in this mailinglist.
and i would say we do not need any external hints, as my version works as it should, and thus we can integrate "its way" into ingos new interface, maybe.

**Comment by ComBin** [comment]
I tried build you branch https://github.com/hoijui/jitsi/tree/systrayNone tray icon not shown. And i try master https://github.com/jitsi/jitsi, icon is displayed but not It does not respond to pressing. :(
I curently test on Kubuntu 16.04, KDE 5.5.5, QT 5.5.1.

**Comment by hoijui** [comment]
ahh.. strange... i though the icon is never shown on that KDE version...
well, on my branch, the icon should not be shown. that is the goal.
what my branch does, is to make sure that if jitsi has no support for a system tray on the current system, the application can still be used in a normal way (there is no way to accidentally close the main window without any way to bring it back, except manually killing the jitsi process and starting jitsi again). i think the main branch has the same goal.
the proper way to make sure that all these systems have a working system tray again with jitsi, is to make sure their new system is supported by the java system tray implementation (part of java/jdk, not jitsi).

**Comment by ComBin** [comment]
So, problem in java/jdk? Is there a bug about this problem in some java bugtracker?

**Comment by hoijui** [comment]
the problem is, as i get it, that many linux systems are changing to a new systray API, and there is no support for that in java (or maybe there are patches already, but they are not deployed yet in any release).

**Comment by hoijui** [comment]
@ibauersachs are you on this, or do you want me to try to get it to work on your new systray system in master?

**Comment by ibauersachs** [comment]
I tried to look at your branch, but it's rather difficult to read as a diff. In my refactorings I tried to avoid to change the UIService interface because it is implemented outside if Jitsi too. And I'm still wondering why the changes I made don't work yet. I won't have time to work on it this week, so if you want to debug this/bring your branch up to date, it would be welcome.

**Comment by ibauersachs** [comment]
Two more notes:
- we need the CLA signed to before I'm able to merge a pull request (https://jitsi.org/icla)
- I wanted to test this library against KDE5: https://github.com/axet/desktop

**Comment by ComBin** [comment]
I have found bug in java bugtracker: http://bugs.java.com/bugdatabase/view_bug.do?bug_id=8146318

**Comment by hoijui** [comment]
nice one @ComBin !
@ibauersachs i already signed the CLA when i submitted patches for otr4j. would i need to sign it again for jitsi?

**Comment by ibauersachs** [comment]
If it was the Atlassian one on Docusign, no. Sorry, I'm not an Atlassian employee and don't have access to the list if signees.

**Comment by hoijui** [comment]
ok.. yes it was that one.
i have it half fixed already.

**Comment by ComBin** [comment]
Hi guys, i tried https://github.com/axet/desktop on my KDE 5.5.5 it's fully work! You may use this lib in Jitsi or get implementation of the Linux system tray from it.

**Comment by hoijui** [comment]
@ibauersachs this one now works for me again (based on your new api stuff):
https://github.com/hoijui/jitsi/tree/systrayNotNew

it also contains some not very related small commits

**Comment by ibauersachs** [comment]
@hoijui I tried your branch on Fedora 23. The problem there is that `java.awt.SystemTray.isSupported()` returns `true` when in fact it isn't (well, there appears a black icon in the lower-left corner, but it's useless). So all your changes (just as mine) are for nought :(
I'll try to integrate axet/desktop now and then we'll see how many changes are still necessary.

Please open a pull request for the non-systray related commits in your branch (i.e. the Netbeans fixes and the comment typo), I'm happy to merge these.

**Comment by hoijui** [comment]
@ibauersachs ahh i see...
that is a problem with fedora/KDE though. i'd say, it makes sense to circumvent this with a hack, something like `isSupported() { if (isFedora && isKde) return false; else ... }`. this hack also makes sense even once axet is an available alternative. my changes also still make sense, because they work as we discussed them in the past, and specifically they work at least on trisquel default setup (and likely others, apart from fedora with KDE). they also make sense, because with a small extension (which is not in this branch yet, but in my original one is), they allow the user to turn off system-tray all together, if he prefers jitsi to work more like a simple GUI application, instead of a quasi-deamon one.

**Comment by ibauersachs** [comment]
@hoijui This is actually not KDE, but Gnome 3. I tried to get the bigger picture about the state of the art of tray icons in various desktop environments and it's a bloody fucking mess.

- Gnome 3: No tray at all. Someone out of their mind decided that persistent notifications should be used instead while clearly not understanding what they're good for.
- Gnome 2: GtkStatusIcon (not sure what that thing uses in the background)
- Unity, KDE: libappindicator
- XFCE: not quite sure, but libappindicator should work, might need the xfce4-indicator-plugin
- LXDE, Cinnamon, Mate, Sugar: no idea.

So my idea right now is:
- Support libappindicator, which should bring proper support for most environments
- If libappindicator is not available and Gnome 3 is running (not sure how to properly detect that yet): no tray
- Fall back to XEmbed otherwise
- Add a new option to disable the tray icon completely, regardless of technical availability
- Create a cmdline-arg to force Jitsi to show if for any reason there's no tray icon and clicking X had hidden the window

(Oh, and I can't really use axet/desktop. It sometimes crashes the JVM.)

Comments?

**Comment by hoijui** [comment]
@ibauersachs wow...
cool you investigated, thanks!

libappindicator support would be done by using this: https://github.com/dorkbox/SystemTray
?

the cmd-line-arg you mentioned might also auto-set the new option to disable the tray icon completely, or at least ask the user if he wants that, in a dialog.

otherwise.. yeah, thanks again for doing that research, and.. looks really messy!
it somehow makes me feel like it would make sense to create a test application that uses the 2 (or more?) implementations of java systray implementations, and does nothing else, and then we log which response we get from which methods (like `isSupported()`) and which exceptions on different systems, and probably we would also have to ask the user whether he actually sees a systray icon (2 or 3 times, like a nagging child ;-) ).
or is that too much work? just do it through jitsi in a kind of long winded trial and error process?

**Comment by ibauersachs** [comment]
Haven't seen that library yet, so thanks for finding it! And I'll also have to thank the creators of that lib for this phrase - I can wholly support them :-)
> Also, screw you gnome-project leads, for making it such a pain-in-the-ass
> to do something so incredibly simple and basic.

I'll try to use this lib in Jitsi and I'll let you know.

**Comment by ibauersachs** [comment]
The dorkbox-lib didn't work out of the box on a Debian 8+Gnome and I'm not sure when I have time again to work on it.

**Comment by bryce-lynch** [comment]
On my system (Arch + Gnome 3.20) whenever I minimize Jitsi, its tray icon becomes a black square and clicking it doesn't do anything - I have to kill the java process and restart it. Given the fact that most of Jitsi's UI bitmaps look too tiny here (could be a dpi issue on my system) I think I'll uninstall it, just wanted to report the issue fyi.

**Comment by Starcommander** [comment]
Hi to all.
I have created a workaround for this issue to reduce the pain.
My patch is using a simple JFrame instead of systemtray.
https://github.com/Starcommander/Jitsi_systray_workaround

**Comment by tjgillies** [comment]
@Starcommander You fix worked for me with jitsi-nightly

**Comment by Starcommander** [comment]
@tjgillies 
Nice that it works for you.
Now I have included appindicator-support for jitsi.
I have tested with success on debian(gnome3) and ubuntu(kde)
Maybe you can also test my changes?

@ibauersachs 
Do you think the changes should be included into jitsi source?

**Comment by ibauersachs** [comment]
I've been working on creating bindings for libappindicator this weekend. Looks good so far and should land in a build in a few days.

**Comment by ibauersachs** [comment]
@Starcommander Sorry for the duplicated effort. I should have mentioned that I've resumed work.

Can you please give build 5541 (once it's out) a try and see how it works for you?

**Comment by Starcommander** [comment]
@ibauersachs 
Hello, and sorry for bad news.
I installed jitsi_2.9.5541-1_amd64.deb, and there is no icon on KDE(ubuntu) or Gnome3(debian).
On ubuntu no error message is shown.
On debian I got that additional exception message on console:

`java.io.IOException: Referenced file does not exist: sc-bundles/yahooaccregwizz.jar
	at org.apache.felix.framework.cache.BundleArchive.createRevisionFromLocation(BundleArchive.java:852)
	at org.apache.felix.framework.cache.BundleArchive.reviseInternal(BundleArchive.java:550)
	at org.apache.felix.framework.cache.BundleArchive.<init>(BundleArchive.java:226)
	at org.apache.felix.framework.cache.BundleCache.getArchives(BundleCache.java:247)
	at org.apache.felix.framework.Felix.init(Felix.java:757)
	at org.apache.felix.framework.Felix.init(Felix.java:626)
	at org.apache.felix.main.Main.main(Main.java:289)
	at net.java.sip.communicator.launcher.SIPCommunicator.main(SIPCommunicator.java:219)
java.io.IOException: Referenced file does not exist: sc-bundles/protocol-yahoo.jar
	at org.apache.felix.framework.cache.BundleArchive.createRevisionFromLocation(BundleArchive.java:852)
	at org.apache.felix.framework.cache.BundleArchive.reviseInternal(BundleArchive.java:550)
	at org.apache.felix.framework.cache.BundleArchive.<init>(BundleArchive.java:226)
	at org.apache.felix.framework.cache.BundleCache.getArchives(BundleCache.java:247)
	at org.apache.felix.framework.Felix.init(Felix.java:757)
	at org.apache.felix.framework.Felix.init(Felix.java:626)
	at org.apache.felix.main.Main.main(Main.java:289)
	at net.java.sip.communicator.launcher.SIPCommunicator.main(SIPCommunicator.java:219)`

Dependencies are installed (libappindicator1 and so on)

Best regards,
Paul

**Comment by ibauersachs** [comment]
Just to make sure, you did actually select AppIndicator as the tray mode in the options? It's set to disabled on all Linux variations by default. And for Gnome 3 it's normal not to see an icon, it doesn't support them (see my comments from mid-2016).

**Comment by Starcommander** [comment]
@ibauersachs 
Options? Do you mean options of linux system, jitsi, maybe java?
On Gnome3 you can locate the AppIndicator Icons on the border (notification) bar by default.
There you can install the "Gnome-Shell-Extensions" for AppIndicator-Support on top-Bar.
In any case AppIndicator icons are working on both Linux systems (I can test it with my own Implemention).
Maybe I have to set options on jitsi to use AppIndicator by default?

**Comment by ibauersachs** [comment]
In Jitsi. There's a new dropdown in Tools->Options->General to select the tray implementation. Regarding Gnome 3, I was referring to a plain installation without extensions.

**Comment by Starcommander** [comment]
@ibauersachs 
Hello again.
Now I got it working, I had to (re)move the jitsi-folder from /home/user/.jitsi/
Because it is difficult to set an option on a Frame, that is not visible :)

Tested on

- debian with Gnome3 --> OK

- ubuntu with KDE --> OK

- fedora 32bit (using Qemu) with LXDE --> AppIndicator works.
But on fedora I got an other exception while closing the login-Frame
(--> frame with SIP, XMPP, and google_talk)

AppIndicator seems to work.
(On existing installations it is possible, that the jitsi-folder makes troubles.)
Thanks for your effort, great job!!!

**Comment by ibauersachs** [comment]
Invisible frame? There's a `--notray` cmdline-option to reset the mode to disabled and show the main UI.

What kind of exception did you get on Fedora with LXDE? Note that basically all distros except the Ubuntu-based ones have a major bug in the libdbusmenu library and print Gtk-ERROR stuff on stdout (or stderr). The bug was fixed in [Ubuntu bug 1203888](https://bugs.launchpad.net/glipper/+bug/1203888). I've reported it to Debian ([bug 850769](http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=850769)), but their Ayatana team has abandoned the project. I doub't it will ever be fixed despite my patch.

I'm closing this issue for now.

**Comment by Starcommander** [comment]
By the way - here the error message, that I got on fedora.
(jitsi crash)
But I fear the output is not very meaningful:

> (java:1508): GLib-GObject-WARNING **: gsignal.c:2523: signal 'child-added' is invalid for instance '0x943392e8' of type 'GtkMenu'
> 
> (java:1508): Gdk-ERROR **: The program 'java' received an X Window System error.
> This probably reflects a bug in the program.
> The error was 'BadWindow (invalid Window parameter)'.
>   (Details: serial 7177 error_code 3 request_code 20 minor_code 0)
>   (Note to programmers: normally, X errors are reported asynchronously;
>    that is, you will receive the error a while after causing it.
>    To debug your program, run it with the --sync command line
>    option to change this behavior. You can then get a meaningful
>    backtrace from your debugger if you break on the gdk_x_error() function.)
> Trace/breakpoint trap (Speicherabzug geschrieben)

I do not think, that has anything to do with AppIndicator, because icon is visible and menu works.

**Comment by ibauersachs** [comment]
The first line about child-added is from AppIndicator via libdbusmenu - the two bugs to which I referred above.
No idea if the second message is a follow-up from that bug. It could also be related to #66 (or #120).
